test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,306 @@
+			t.Logf("Failed to delete externalnode, err: %v", err)
+		}
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, false)
+		verifyUpLinkAfterCleanup(vm)
+	}
+}
+
+func verifyExternalEntityExistence(t *testing.T, data *TestData, eeName string, vmNodeName string, expectExists bool) {
+	if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+		t.Logf("Verifying ExternalEntity %s, expectExists %t", eeName, expectExists)
+		_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nameSpace).Get(context.TODO(), eeName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			t.Errorf("Failed to get ExternalEntity %s by ExternalNode %s: %v", eeName, vmNodeName, err)
+			return false, err
+		}
+
+		if expectExists {
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		}
+		// ExternalEntity is expected to be deleted, check for error
+		if err != nil {
+			return true, nil
+		}
+		return false, nil
+
+	}); err != nil {
+		op := "created"
+		if !expectExists {
+			op = "deleted"
+		}
+		assert.NoError(t, err, "Failed to verify ExternalEntity %s %s by ExternalNode %s", eeName, op, vmNodeName)
+	}
+}
+
+func testExternalNode(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyExternalNodeRealization := func(vm vmInfo) {
+		if err := wait.PollImmediate(30*time.Second, 1*time.Minute, func() (done bool, err error) {
+			t.Logf("Verify host interface configuration for VM: %s", vm.nodeName)
+			exists, err := verifyInterfaceIsInOVS(t, data, vm)
+			return exists, err
+		}); err != nil {
+			assert.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
+			return
+		}
+
+		var tempVm vmInfo
+		if vm.osType == "Windows" {
+			tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+		} else {
+			tempVm = getVMInfo(t, data, vm.nodeName)
+		}
+		assert.Exactly(t, vm.ifName, tempVm.ifName, "Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+		assert.Exactly(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
+	}
+	for _, vm := range vmList {
+		hash := sha1.New() // #nosec G401: not used for security purposes
+		io.WriteString(hash, vm.ifName)
+		hashedIfName := hex.EncodeToString(hash.Sum(nil))
+		// Generate ExternalEntity name from ExternalNode
+		eeName := vm.nodeName + "-" + hashedIfName[:interfaceNameLength]
+		t.Logf("Running verifyExternalEntityExistence")
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, true)
+		t.Logf("Running verifyExternalNodeRealization")
+		verifyExternalNodeRealization(vm)
+	}
+}
+
+func getVMInfo(t *testing.T, data *TestData, nodeName string) (info vmInfo) {
+	var vm vmInfo
+	vm.nodeName = nodeName
+	var cmd string
+	cmd = "ip -o -4 route show to default | awk '{print $5}'"
+	vm.osType = "Linux"
+	rc, ifName, _, err := data.RunCommandOnNode(nodeName, cmd)
+	require.Nil(t, err, "Failed to run command %s on VM %s, err %v, rc %d", cmd, nodeName, err, rc)
```

Comment:
use `require.NoError` for errors, not `require.Nil`. It makes the intent more explicit and in case of an error, the test output is more helpful

Reply:
Sure. Updated it.